### PR TITLE
adding changes for adv query support in elastalert

### PR DESCRIPTION
**Pull request by sivatarunp**
## Description

This PR supports adv query in elastalert. The idea is to give the adv query which is provided directly in kibana using filter string of elastalert, and evaluate the final metric using thresholds.
Currently adv query has limited output data formats, of which aggregations are of our prime concern because, adv query in alerts needs some final metric calculation, which falls under aggregations output type of advance query.

This rule type now takes given query, evaluates the query for a buffer_time window and compares the evaluated output (field of comparison is to be given in alert definition) with thresholds and fires the alert.

Since most  of alerts here would be on some stat, MetricAggregation way of solving the problem is taken here 

## Checklist

- [ ] I have reviewed the [contributing guidelines](https://github.com/jertel/elastalert2/blob/master/CONTRIBUTING.md).
- [ ] I have included unit tests for my changes or additions.
- [ ] I have successfully run `make test-docker` with my changes.
- [ ] I have manually tested all relevant modes of the change in this PR.
- [ ] I have updated the [documentation](https://elastalert2.readthedocs.io).
- [ ] I have updated the [changelog](https://github.com/jertel/elastalert2/blob/master/CHANGELOG.md).

### CODE REVIEW

File: elastalert/ruletypes.py [hunk 1]
```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
                 placeholder.update({self.rules['query_key']: qk})
             self.handle_event(placeholder, 0, qk)
 
+class AdvQueryRule(RuleType):
+    """ A rule that uses a query_string query to perform a search """
```
Comment (RashmiRam) [inline elastalert/ruletypes.py]: Change comment to include that this is specific for advanced query

File: elastalert/loaders.py [hunk 2]
```diff
@@ -93,7 +93,8 @@ class RulesLoader(object):
         'metric_aggregation': ruletypes.MetricAggregationRule,
         'percentage_match': ruletypes.PercentageMatchRule,
         'spike_aggregation': ruletypes.SpikeMetricAggregationRule,
-        'error_rate': ruletypes.ErrorRateRule  #Adding Error Rate Rule type
+        'error_rate': ruletypes.ErrorRateRule,  #Adding Error Rate Rule type
+        'adv_query': ruletypes.AdvQueryRule
```
Comment (RashmiRam) [inline elastalert/loaders.py]: Name it properly only.

File: elastalert/elastalert.py [hunk 3]
```diff
@@ -747,7 +775,12 @@ def run_query(self, rule, start=None, end=None, scroll=False):
         elif isinstance(rule_inst, ErrorRateRule):
             data = self.get_error_rate(rule, start, end)
         elif rule.get('aggregation_query_element'):
-            data = self.get_hits_aggregation(rule, start, end, index, rule.get('query_key', None))
+            print("in agg query element")
```
Comment (RashmiRam) [inline elastalert/elastalert.py]: Use logger.

File: elastalert/elastalert.py [hunk 4]
```diff
@@ -624,6 +625,33 @@ def get_hits_aggregation(self, rule, starttime, endtime, index, query_key, term_
...
+        base_query = self.get_query(
+            rule_filter,
+            starttime,
+            endtime,
+            timestamp_field=rule['timestamp_field'],
+            sort=False,
+            to_ts_func=rule['dt_to_ts'],
+        )
+        request = get_msearch_query(base_query,rule)
+        print("yoyo")
```
Comment (RashmiRam) [inline elastalert/elastalert.py]: cleanup print before merging

Comment (RashmiRam) [reply elastalert/elastalert.py]: How is this different from other hits aggregation method call? why separate `get_adv_query_aggregation` ?

File: elastalert/ruletypes.py [hunk 5]
```diff
@@ -301,6 +301,7 @@ def check_for_match(self, key, end=False):
             event = self.occurrences[key].data[-1][0]
             if self.attach_related:
                 event['related_events'] = [data[0] for data in self.occurrences[key].data[:-1]]
+            event['count'] = self.occurrences[key].data[-1][1]
```
Comment (RashmiRam) [inline elastalert/ruletypes.py]: why is this required in frequency rule type?

File: elastalert/ruletypes.py [hunk 6]
```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
...
+                            if self.rules['alert_field'] in k:
+                                if self.crossed_thresholds(v['value']):
+                                    match={"key": key,"value":v['value'],"key_value":data_key_value,self.rules['timestamp_field']: timestamp}
+                                    self.add_match(match)
+                data_key_value = local_data_key
+    
+    def crossed_thresholds(self, metric_value):
+        if metric_value is None:
+            return False
+        if 'max_threshold' in self.rules and float(metric_value) > self.rules['max_threshold']:
```
Comment (RashmiRam) [inline elastalert/ruletypes.py:785]: what if they provided both max and min? We should check between. Isn't it?

File: elastalert/ruletypes.py [hunk 7]
```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
...
+        if len(data_value) > 0:
+            for data in data_value:
+                local_data_key = data_key_value
+                data_key_value = data_key_value+ ','+ data['key'] if data_key_value else data['key']
+                for k,v in data.items():
+                    if k!= 'key' and k!= 'doc_count':
+                        if 'buckets' in v:
+                            local_key = key
+                            key = key + ','+k
+                            self.check_matches_recursive(key,v['buckets'],data_key_value)
```
Comment (RashmiRam) [inline elastalert/ruletypes.py]: where are we passing the `timestamp` param here?

File: elastalert/ruletypes.py [hunk 8]
```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
...
+                if self.crossed_thresholds(value['value']):
+                    match={"key":key,"value":value['value']}
+                    self.add_match(match)
+
+    def check_matches_recursive(self,data_key,data_value,timestamp,data_key_value=''):
+        key = data_key
+        if len(data_value) > 0:
+            for data in data_value:
+                local_data_key = data_key_value
+                data_key_value = data_key_value+ ','+ data['key'] if data_key_value else data['key']
```
Comment (RashmiRam) [inline elastalert/ruletypes.py]: this data_key_value name is confusing.

File: elastalert/ruletypes.py [hunk 9]
```diff
@@ -720,6 +721,65 @@ def garbage_collect(self, ts):
...
+                if len(value['buckets']) >=0 :
+                    self.check_matches_recursive(key,value['buckets'],timestamp)
+            else:
+                if self.crossed_thresholds(value['value']):
+                    match={"key":key,"value":value['value']}
+                    self.add_match(match)
+
+    def check_matches_recursive(self,data_key,data_value,timestamp,data_key_value=''):
+        key = data_key
+        if len(data_value) > 0:
```
Comment (RashmiRam) [inline elastalert/ruletypes.py]: Instead of combining the check_matches & finding key in nested dic logic in same method, Can we have separate method to just do recursive find of a key in nested dict and invoke it by passing the alert_field key and response object?

**Review by RashmiRam — CHANGES_REQUESTED** [review]

File: elastalert/ruletypes.py (hunk 9 above)
Comment (sivatarunp) [reply elastalert/ruletypes.py]: Can do that Rashmi. Generally in check_matches function only thresholds are compared in all rule types. Hence followed the same logic here. Even in Metric aggregation type, elastalert does the same

File: elastalert/ruletypes.py (hunk 8 above)
Comment (sivatarunp) [reply elastalert/ruletypes.py]: changed it

File: elastalert/ruletypes.py (hunk 7 above)
Comment (sivatarunp) [reply elastalert/ruletypes.py]: Rectified

File: elastalert/ruletypes.py (hunk 6 above)
Comment (sivatarunp) [reply elastalert/ruletypes.py:785]: Currently alert conditions are only if greater than something, or less than something. There is no in between part. If  we  want to support it we can add

File: elastalert/ruletypes.py (hunk 5 above)
Comment (sivatarunp) [reply elastalert/ruletypes.py]: That was from different PR, by mistake got added here. Removed it.

File: elastalert/elastalert.py (hunk 4 above)
Comment (sivatarunp) [reply elastalert/elastalert.py]: other aggregation query will build aggs query based on the params (which aggregation to be built)and query_key fields. We dont require to build aggs query here, direct query string query itself will work. SO it will be similar to get_hits methods (called for Any type of rule). But with aggregations in output.

File: elastalert/elastalert.py (hunk 3 above)
Comment (sivatarunp) [reply elastalert/elastalert.py]: Done

File: elastalert/loaders.py (hunk 2 above)
Comment (sivatarunp) [reply elastalert/loaders.py]: Done

File: elastalert/ruletypes.py (hunk 1 above)
Comment (sivatarunp) [reply elastalert/ruletypes.py]: DOne

File: elastalert/ruletypes.py (hunk 6 above)
Comment (aravind-musigumpula) [reply elastalert/ruletypes.py:776]: Users can provide both max and min thresholds in the same rule, and this will work for it.
@RashmiRam is this what you are asking?

Comment (RashmiRam) [reply elastalert/ruletypes.py:785]: Giving alert arguments like max/min threshold & not providing between check seems counter intuitive.

File: elastalert/loaders.py [hunk 10]
```diff
@@ -93,7 +93,8 @@ class RulesLoader(object):
         'metric_aggregation': ruletypes.MetricAggregationRule,
         'percentage_match': ruletypes.PercentageMatchRule,
         'spike_aggregation': ruletypes.SpikeMetricAggregationRule,
-        'error_rate': ruletypes.ErrorRateRule  #Adding Error Rate Rule type
+        'error_rate': ruletypes.ErrorRateRule,  #Adding Error Rate Rule type
+        'advance_search': ruletypes.AdvanceSearchRule
```
Comment (RashmiRam) [inline elastalert/loaders.py]: no better name than this?

Comment (RashmiRam) [reply elastalert/loaders.py]: If no better name, at least change it to `advanced_query`? We exposed this feature as advanced query only?

File: elastalert/ruletypes.py (hunk 9 above)
Comment (RashmiRam) [reply elastalert/ruletypes.py]: I understand what `check_matches` does. But what I meant is that, we can have separate method to do this nested parsing [https://stackoverflow.com/a/29652561](https://stackoverflow.com/a/29652561) and make use of it.

**Review by RashmiRam — CHANGES_REQUESTED** [review]
